Add Zipkin tests to testbed

mat-rumian opened the pull request:
**Description:** 
I've added Zipkin data receiver and sender for testbed. Few tests were updated and two new created.

**Link to tracking Issue:**
[Issue-482](https://github.com/open-telemetry/opentelemetry-collector/issues/482)

**Testing:** 
- TestGeneratorAndBackend -> changed the test name, Jaeger and Zipkin as test cases
- TraceNoBackend10kSPS -> created test scenario, possibility to add new test cases
- Trace10kSPS -> added Zipkin test case

**Documentation:**
No new docs.

mat-rumian commented:
I signed it

mat-rumian commented on testbed/tests/trace_test.go:
```diff
@@ -65,6 +65,15 @@ func TestTrace10kSPS(t *testing.T) {
+			testbed.NewZipkinDataSender(testbed.GetAvailablePort(t)),
+			testbed.NewZipkinDataReceiver(testbed.GetAvailablePort(t)),
+			testbed.ResourceSpec{
+				ExpectedMaxCPU: 42,
```
The values for `ExpectedMaxCPU` and `ExpectedMaxRAM` are the things which I'm not sure.

tigrannajaryan replied:
The values need to be determined after running on TravisCI. We use (max-observed * 1.3) formula for limits.

mat-rumian commented on testbed/tests/trace_test.go:
```diff
@@ -126,6 +135,52 @@ func TestTraceNoBackend10kSPSJaeger(t *testing.T) {
+			defer tc.Stop()
+
+			tc.SetResourceLimits(testbed.ResourceSpec{
+				ExpectedMaxCPU: 60,
```
Same as above comment.

tigrannajaryan commented on testbed/testbed/senders.go:
```diff
@@ -68,6 +69,7 @@ type MetricDataSender interface {
 // DataSenderOverTraceExporter partially implements TraceDataSender via a TraceExporter.
 type DataSenderOverTraceExporter struct {
 	exporter exporter.TraceExporter
+	Address  string
```
Is this used anywhere?

mat-rumian replied:
No, sorry I was thinking to use it instead of `Port` and later I forgot to delete it.

tigrannajaryan commented on testbed/tests/testdata/agent-config.yaml:
```diff
@@ -5,22 +5,26 @@ receivers:
   pipelines:
     traces:
-      receivers: [jaeger]
+      receivers: [jaeger,zipkin]
```
Same here. No need to enable zipkin in all tests.

tigrannajaryan commented on testbed/tests/testdata/memory-limiter.yaml:
```diff
@@ -5,13 +5,16 @@ receivers:
   logging:
     loglevel: info
-
+  zipkin:
```
and here.

tigrannajaryan commented on testbed/tests/trace_test.go:
```diff
@@ -126,6 +135,52 @@ func TestTraceNoBackend10kSPSJaeger(t *testing.T) {
 	}
 }
 
+func TestTraceNoBackend10kSPSZipkin(t *testing.T) {
```
This is a duplicate of TestTraceNoBackend10kSPSJaeger. Instead of duplicating code it is better to add another looping dimension to TestTraceNoBackend10kSPSJaeger, so that it performs the same scenarios using different receivers/exporters (and rename TestTraceNoBackend10kSPSJaeger to TestTraceNoBackend10kSPS).

mat-rumian replied:
@tigrannajaryan so the question is, should I leave this tests as it was (because it needs also specific config yaml file with defined Zipkin receiver) or add as test case zipkin receiver and create two new config yamls?

tigrannajaryan commented on testbed/tests/trace_test.go:
```diff
@@ -230,10 +285,10 @@ func TestTraceBallast1kSPSAddAttrs(t *testing.T) {
 // and received by MockBackend and calls user-supplied verification function on
 // received span.
-func verifySingleSpan(
+func verifySingleJaegerSpan(
```
This is a generic function that is independent from the format. It is not Jaeger-specific. The purpose of changing the name of the function is not clear.

tigrannajaryan commented on testbed/tests/trace_test.go:
```diff
@@ -271,10 +326,10 @@ func verifySingleSpan(
 }
 
-func TestTraceAttributesProcessor(t *testing.T) {
+func TestJaegerTraceAttributesProcessor(t *testing.T) {
```
This change is against the intent of the test. The test is already intended to test different formats using subtests. Please rebase and you will see that now it has 2 subtests, each using different senders and receiver. Zipkin can be added to as another subtest.

tigrannajaryan commented on testbed/tests/trace_test.go:
```diff
@@ -298,14 +353,14 @@ func TestTraceAttributesProcessor(t *testing.T) {
+   actions:
+     - action: insert
+       key: "new_attr"
+       value: "string value"
```
Why this change?

mat-rumian replied:
It was just a mistake space, sorry.

tigrannajaryan commented on testbed/tests/trace_test.go:
```diff
@@ -349,16 +405,139 @@ func TestTraceAttributesProcessor(t *testing.T) {
+// verifySingleZipkinSpan sends a single span to Collector, waits until the span is forwarded
+// and received by MockBackend and calls user-supplied verification function on
+// received span.
+func verifySingleZipkinSpan(
```
What is the difference of this function from verifySingleSpan? Can we use the same verification function for Zipkin?

tigrannajaryan commented on testbed/tests/trace_test.go:
```diff
@@ -349,16 +405,139 @@ func TestTraceAttributesProcessor(t *testing.T) {
+	assert.EqualValues(t, 1, count, "must receive one Zipkin span")
+}
+
+func TestZipkinTraceAttributesProcessor(t *testing.T) {
```
This should be possible to add to TestTraceAttributesProcessor as a subtest.

mat-rumian replied:
@tigrannajaryan I've created a separated test because IMHO processor used in `TestTraceAttributesProcessor` was not working correctly for Zipkin. When I removed from include attributes the `services` key then the test was modifying traces and they were passing. Maybe I'm missing something about data models or something else?

tigrannajaryan commented on testbed/testbed/mock_backend_test.go:
```diff
@@ -49,6 +49,35 @@ func TestGeneratorAndBackend(t *testing.T) {
 	assert.Equal(t, lg.DataItemsSent(), mb.DataItemsReceived())
 }
 
+func TestZipkinGeneratorAndBackend(t *testing.T) {
```
Instead of creating this func by copying TestGeneratorAndBackend it would be nicer to change TestGeneratorAndBackend to use subtests and make Jaeger and Zipkin the 2 subtests.

mat-rumian commented on testbed/tests/testdata/add-attributes-config.yaml:
```diff
@@ -31,6 +36,6 @@ processors:
+      receivers: [jaeger,zipkin]
       processors: [attributes,queued_retry]
-      exporters: [opencensus]
+      exporters: [opencensus,zipkin]
```
Thanks for this knowledge. I will revert the changes in yaml files and will be happy to take care of new matrix test.

bogdandrutu commented:
Please rebase

tigrannajaryan commented on testbed/tests/trace_test.go:
```diff
@@ -89,49 +99,86 @@ func TestTrace10kSPS(t *testing.T) {
-
-			tc.Sleep(tc.Duration)
+			},
+			[]testbed.ProcConfiguration{
+				{
+					Name:                "NoMemoryLimit",
+					Processor:           noLimitProcessors,
+					ExpectedMaxRAM:      200,
+					ExpectedMinFinalRAM: 100,
+				},
+				{
+					Name:                "MemoryLimit",
+					Processor:           limitProcessors,
+					ExpectedMaxRAM:      60,
+					ExpectedMinFinalRAM: 10,
+				},
+			},
+		},
```
This seems to be the same for Jaeger and Zipkin, so would be nice to extract as a var and reuse.

tigrannajaryan commented on testbed/testbed/utils.go:
```diff
@@ -23,3 +23,10 @@ import (
 	return int(testutils.GetAvailablePort(t))
 }
+
+type ProcConfiguration struct {
```
`Proc` is a bit unclear abbreviation. Let's use `Processor`. `Configuration` can be safely abbreviated to `Config`, as it is commonly done in the codebase, so `ProcessorConfig` should be fine.

tigrannajaryan replied:
Also add a common on what this struct is for.

tigrannajaryan commented on testbed/testbed/utils.go:
```diff
@@ -23,3 +23,10 @@ import (
+
+type ProcConfiguration struct {
+	Name                string
+	Processor           map[string]string
```
```suggestion
	// map of processor types to their config YAML to use.
	ProcessorConfig           map[string]string
```

tigrannajaryan commented:
Please raise the test limits to avoid build failure.

tigrannajaryan commented on testbed/testbed/utils.go:
```diff
@@ -23,3 +23,13 @@ import (
 	return int(testutils.GetAvailablePort(t))
 }
+
+// Structure used for TestTraceNoBackend10kSPS.
+// Defines RAM usage range for defined processor type.
+type ProcessorConfig struct {
+	Name string
+	// map of processor types to their config YAML to use.
+	ProcessorConfig     map[string]string
+	ExpectedMaxRAM      uint32
+	ExpectedMinFinalRAM uint32
+}
```
This is used by code in `testbed/tests`. Please move this closer to where it is needed, perhaps close to `ScenarioTestTraceNoBackend10kSPS` which uses it (or in a separate .go file if it needs to be used by many functions from different files in that package). There does not seem to be a need to declare this in the testbed at the moment.

mat-rumian commented:
@tigrannajaryan 

> Please raise the test limits to avoid build failure.

I'm not sure if rising the limits will be enough. Please take a look in the logs from the tests:

```2020/04/06 10:29:55 Starting load generator at 10000 items/sec.
2020/04/06 10:29:56 Agent RAM (RES):   0 MiB, CPU: 0.0%, Sent: 9620 items, Received:    0 items
2020/04/06 10:29:57 Agent RAM (RES):   0 MiB, CPU: 0.0%, Sent:19380 items, Received:16400 items
2020/04/06 10:29:58 Agent RAM (RES):   0 MiB, CPU: 0.0%, Sent:28850 items, Received:24600 items
2020/04/06 10:29:58 Performance error: CPU consumption is 64.7%, max expected is 42%
2020/04/06 10:29:58 Gracefully terminating Agent pid=12479, sending SIGTEM...
2020/04/06 10:29:58 Cannot send traces: Post "http://localhost:44279/api/v2/spans": dial tcp 127.0.0.1:44279: connect: connection refused
2020/04/06 10:29:59 Agent process stopped, exit code=0
2020/04/06 10:29:59 CPU consumption is 64.7%, max expected is 42%
    TestTrace10kSPS/Zipkin: test_case.go:364: CPU consumption is 64.7%, max expected is 42%
2020/04/06 10:29:59 Stopped generator. Sent:29930 items
```

I think this `Cannot send traces: Post "http://localhost:44279/api/v2/spans": dial tcp 127.0.0.1:44279: connect: connection refused` could cause a CPU consumption and looks like not all traces were sent. I was trying to investigate it but I didn't find a way to solve it.

tigrannajaryan commented:
"Cannot send traces" may be a result of termination because of exceeding CPU limits.

tigrannajaryan commented:
Try increasing CPU limits and rerun to see if it passes. If it passes we are good. If it fails again with "Cannot send traces" without hitting resource limits then we can have a look.

mat-rumian commented:
@tigrannajaryan right now two tests are failing:
* TestTraceNoBackend10kSPS for NoLimitMemory - RSS (around 28MB) is lower than ExpectedMinFinalRAM (100MB)
* TraceAttributesProcessor/Zipkin - as I mentioned few comments above - Zipkin test requires some changes in the body. I think TraceAttributesProcessor should be split in small single tests.

tigrannajaryan commented:
@mat-rumian sorry for the delayed reply.

>TestTraceNoBackend10kSPS for NoLimitMemory - RSS (around 28MB) is lower than ExpectedMinFinalRAM (100MB)

This likely means the test or the agent are not doing what is expected. Since there is no backend the memory should grow until that minimum because that's how much space generated traces are expected to use in the memory

tigrannajaryan commented:
>TraceAttributesProcessor/Zipkin - as I mentioned few comments above - Zipkin test requires some changes in the body. I think TraceAttributesProcessor should be split in small single tests.

To avoid further delays and make a progress on this PR I would advise to limit the scope of this PR to what already works. You can work on attributes test in a follow up PR if you want.

mat-rumian commented:
@tigrannajaryan I've added missing processors to `TestTraceNoBackend10kSPS` and everything looks fine now. I've also removed Zipkin tc from `TraceAttributesProcessor`. I will create an issue about it. I think PR is ready to be merged :)

tigrannajaryan commented on testbed/tests/scenarios.go:
```diff
@@ -235,3 +237,48 @@ func Scenario1kSPSWithAttrs(t *testing.T, args []string, tests []TestCase, opts
+
+// Structure used for TestTraceNoBackend10kSPS.
+// Defines RAM usage range for defined processor type.
+type ProcessorConfig struct {
```
From what I see this is only used in this package. Let's make this unexported `processorConfig` since it does not need to be exported.

tigrannajaryan commented on testbed/tests/scenarios.go:
```diff
@@ -235,3 +237,48 @@ func Scenario1kSPSWithAttrs(t *testing.T, args []string, tests []TestCase, opts
+type ProcessorConfig struct {
+	Name string
+	// map of processor types to their config YAML to use.
+	ProcessorConfig     map[string]string
```
It is a bit confusing when the name of the field is the same as the struct it is in. Can you come up with a better name?

mat-rumian commented:
Done 💯
